src/Markdown/Parser.elm

```diff
@@ -644,71 +641,88 @@ parseAllInlinesHelp state rawBlocks parsedBlocks =
+                                {linkReferenceDefinitions=[], rawBlocks= []}
 
             ( OpenBlockOrParagraph (UnparsedInlines body1), (OpenBlockOrParagraph (UnparsedInlines body2)) :: rest ) ->
-                OpenBlockOrParagraph (UnparsedInlines (joinRawStringsWith "\n" body2 body1))
-                    :: rest
+                { linkReferenceDefinitions = state.linkReferenceDefinitions
+                , rawBlocks = OpenBlockOrParagraph (UnparsedInlines (joinRawStringsWith "\n" body2 body1))
+                    :: rest}
 
             ( SetextLine LevelOne _, (OpenBlockOrParagraph unparsedInlines) :: rest ) ->
-                Heading 1 unparsedInlines
-                    :: rest
+                { linkReferenceDefinitions = state.linkReferenceDefinitions
+                , rawBlocks =Heading 1 unparsedInlines
+                    :: rest}
 
             ( SetextLine LevelTwo _, (OpenBlockOrParagraph unparsedInlines) :: rest ) ->
-                Heading 2 unparsedInlines
-                    :: rest
+                { linkReferenceDefinitions = state.linkReferenceDefinitions
+                , rawBlocks =Heading 2 unparsedInlines
+                    :: rest}
 
             ( TableDelimiter (Markdown.Table.TableDelimiterRow text alignments), (OpenBlockOrParagraph (UnparsedInlines rawHeaders)) :: rest ) ->
                 case TableParser.parseHeader (Markdown.Table.TableDelimiterRow text alignments) rawHeaders of
                     Ok (Markdown.Table.TableHeader headers) ->
-                        Table (Markdown.Table.Table headers []) :: rest
+                        { linkReferenceDefinitions = state.linkReferenceDefinitions
+                        , rawBlocks =Table (Markdown.Table.Table headers []) :: rest}
 
                     Err _ ->
-                        OpenBlockOrParagraph (UnparsedInlines (joinRawStringsWith "\n" rawHeaders text.raw))
-                            :: rest
+                        { linkReferenceDefinitions = state.linkReferenceDefinitions
+                        , rawBlocks =OpenBlockOrParagraph (UnparsedInlines (joinRawStringsWith "\n" rawHeaders text.raw))
+                            :: rest}
 
             ( Table updatedTable, (Table _) :: rest ) ->
-                Table updatedTable :: rest
+                { linkReferenceDefinitions = state.linkReferenceDefinitions
+                    , rawBlocks = Table updatedTable :: rest}
 
             _ ->
-                newRawBlock :: state.rawBlocks
-    }
-
-
+                { linkReferenceDefinitions = state.linkReferenceDefinitions
+                    , rawBlocks = newRawBlock :: state.rawBlocks}
 
 -- RAW BLOCK PARSER
-
-
 stepRawBlock : State -> Parser (Step State State)
 stepRawBlock revStmts =
     -- Some blocks can't immediately follow a body
     oneOf
         [ Helpers.endOfFile
-            |> map (\_ -> Done revStmts)
+            |> map (\_ -> Done (completeBlocks revStmts))
```

Comment:
If you change this to `andThen`, I believe you could use `succeed` and `fail` to pass through the parsed value or error. That would allow you to resolve your todo comment `-- TODO return this error`.